File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
def dfs(numbers, target, total, depth):
    if depth == len(numbers):
        if total == target:
            return 1
        else:
            return 0
    else:
        answer = 0
        answer += dfs(numbers, target, total + numbers[depth], depth + 1)
        answer += dfs(numbers, target, total - numbers[depth], depth + 1)
        return answer
        
            
def solution(numbers, target):
    answer = dfs(numbers, target, 0, 0)
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;
class Solution {
    static int N;
    static int answer;
    public int solution(int[] numbers, int target) {
        answer = 0;
        N = numbers.length;
        dfs(0, numbers, target, 0);
        return answer;
    }
    
    static public void dfs(int depth, int[] numbers, int target, int now) {
        if (depth == N) {
            if (now == target) {
                answer++;
            }
            return;
        }
        dfs(depth + 1, numbers, target, now + numbers[depth]);
        dfs(depth + 1, numbers, target, now - numbers[depth]);
    }
}